crates/subspace-proof-of-time/src/pot_aes.rs

```diff
@@ -0,0 +1,125 @@
+//! AES related functionality.
+
+use aes::cipher::generic_array::GenericArray;
+use aes::cipher::{BlockEncrypt, KeyInit};
+use aes::Aes128;
+use subspace_core_primitives::{PotBytes, PotCheckpoint, PotKey, PotSeed};
+
+pub(crate) fn create(
+    seed: &PotSeed,
+    key: &PotKey,
+    num_checkpoints: u16,
+    checkpoint_iterations: u64,
+) -> Vec<PotCheckpoint> {
+    let key = GenericArray::from(PotBytes::from(*key));
+    let cipher = Aes128::new(&key);
+    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
+
+    let mut checkpoints = Vec::with_capacity(num_checkpoints as usize);
+    for _ in 0..num_checkpoints {
+        for _ in 0..checkpoint_iterations {
+            // Encrypt in place to produce the next block.
+            cipher.encrypt_block(&mut cur_block);
+        }
+        let bytes: PotBytes = cur_block.into();
+        checkpoints.push(PotCheckpoint::from(bytes));
+    }
+    checkpoints
+}
+
+pub(crate) fn verify(
+    seed: &PotSeed,
+    key: &PotKey,
+    checkpoints: &[PotCheckpoint],
+    checkpoint_iterations: u64,
+) -> bool {
+    let key = GenericArray::from(PotBytes::from(*key));
+    let cipher = Aes128::new(&key);
+    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
+
+    for checkpoint in checkpoints {
```

Comment:
This is not how it's described in the spec. The verifier should be able to a) verify any pair of checkpoints independently and b) do so in parallel. If you plan to add parallelization later, at least leave a TODO. It is critical to have verification parallelizable.

Reply:
https://github.com/subspace/subspace/pull/1584/files#diff-0f3fac90f4c5bcd830f024a86381a4e3cbdcb59ba6def33ea214ef172a8e7bab (the previous PR) has this in parallel. Removed it as it may not be no_std compliant, will add TODO to bring it back

> The verifier should be able to a) verify any pair of checkpoints independently

I wanted to avoid repeated init of the key, which is why it is structured this way. Also, when we switch to parallel, we would need all checkpoints one shot. Pls update the spec if possible to take the array as input, instead of individual pairs

Reply:
also, this is symmetric with the create(), which produces the array of checkpoints

Reply:
The whole point was that verification be parallel though.
What do you mean?
>Pls update the spec if possible to take the array as input, instead of individual pairs
The spec has 2 verification functions defined: parallelizable [one](https://www.notion.so/subspacelabs/PoT-Specification-WIP-4e17a6d5b03a4abea864ba2d0b97970e?pvs=4#0556b3991ff44b78af48cc67bcd92592) that takes all checkpoints, and simple [one](https://www.notion.so/subspacelabs/PoT-Specification-WIP-4e17a6d5b03a4abea864ba2d0b97970e?pvs=4#4f608025ba1b449896e221dadd0bcbe1) that takes a pair.

Reply:
We can proceed with the current verification, but at least add a TODO for parallel. That is critical.

Reply:
Parallel can be done in `no_std` environment (technically), so `no_std` and parallelism are not conflicting. It might be the case that our implementation will require `std`, but it is not an issue. Parallel implementation can be a feature-flag, IIRC `rayon` will still work with one thread in `no_std` environments.

Reply:
> The whole point was that verification be parallel though.
What do you mean?

I understand why it is important, I was trying to way the code and spec  need to be 1:1 (it is hard to code as is in the spec). I put back the parallel verification(which is no_std compliant as well). Benchmarking would need to determine the exact params (num checkpoints, checkpoint params) to occupy 1 sec.

Reply:
Implementation almost never matches the spec 100% word to word in a sense that there are other considerations in implementation that originate from performance and other aspects of the final software. It needs to do exactly what spec says, but the way it does it can vary.